proteuslib/property_models/tests/property_test_harness.py

```diff
@@ -0,0 +1,368 @@
+        m = frame_stateblock
+        blk = m.fs.stream[0]
+        stats = m._test_objs.stateblock_statistics
+        if number_variables(blk) != stats['number_variables']:
+            raise Exception(
+                "The number of variables were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_variables(blk),
+                    num_test=stats['number_variables']))
+        if number_total_constraints(blk) != stats['number_total_constraints']:
+            raise Exception(
+                "The number of constraints were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_total_constraints(blk),
+                    num_test=stats['number_total_constraints']))
+        if number_unused_variables(blk) != stats['number_unused_variables']:
+            raise Exception(
+                "The number of unused variables were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_unused_variables(blk),
+                    num_test=stats['number_unused_variables']))
+        if degrees_of_freedom(blk) != stats['default_degrees_of_freedom']:
+            raise Exception(
+                "The number of degrees of freedom were {num}, but {num_test} was "
+                "expected ".format(
+                    num=degrees_of_freedom(blk),
+                    num_test=stats['default_degrees_of_freedom']))
+
+    @pytest.mark.unit
+    def test_units_consistent(self, frame_stateblock):
+        m = frame_stateblock
+        assert_units_consistent(m)
+
+    @pytest.mark.unit
+    def test_scaling(self, frame_stateblock):
+        m = frame_stateblock
+        calculate_scaling_factors(m.fs.stream[0])
+
+        # check that all variables have scaling factors
+        unscaled_var_list = list(unscaled_variables_generator(m.fs.stream[0]))
+        if len(unscaled_var_list) != 0:
+            unscaled_var_name_list = [v.name for v in unscaled_var_list]
+            raise Exception(
+                "The following variable(s) are unscaled: {lst}".format(
+                    lst=unscaled_var_name_list))
+
+        # check that all constraints have been scaled
+        unscaled_constraint_list = list(unscaled_constraints_generator(m.fs.stream[0]))
+        if len(unscaled_constraint_list) != 0:
+            unscaled_constraint_name_list = [c.name for c in unscaled_constraint_list]
+            raise Exception(
+                "The following constraint(s) are unscaled: {lst}".format(
+                    lst=unscaled_constraint_name_list))
+
+    @pytest.mark.component
+    def test_default_initialization(self, frame_stateblock):
+        m = frame_stateblock
+
+        # initialize
+        m.fs.stream.initialize()
+
+        # check results
+        for (v_name, ind), val in m._test_objs.default_solution.items():
+            var = getattr(m.fs.stream[0], v_name)
```

Comment:
Given that initialize has no way of returning solver status yet, it might also be worth adding a check for the value (i.e. residual) of all constraints.

Reply:
Added a function based on the infeasibility tool in pyomo